### ISSUE TITLE
`oc logs -f bc/x` sporadically returns partial logs during active build

### ISSUE THREAD

**Issue by jupierce** [openshift/origin#12160]
Running `oc logs -f bc/x` on a running build will periodically fail to return all logs back to the console. 

##### Version
oc v1.5.0-alpha.0+0c89011-92
kubernetes v1.4.0+776c994
features: Basic-Auth GSSAPI Kerberos SPNEGO

docker version: Docker version 1.10.3, build f476348/1.10.3
journald version: systemd 229

##### Steps To Reproduce
1. Run https://gist.github.com/jupierce/d877e02c478c1c3127d73aef38368b6a
2. The script will periodically exit with the error "Did not receive full logs"

##### Current Result
Sporadic failures from the script. One variation of which is captured in #12142.

##### Expected Result
If #12142 is also fixed, this script should run indefinitely.

**Comment by jupierce** [comment]
fyi @bparees

**Comment by jupierce** [comment]
I've now seen three variations:
**I've not been able to reproduce any of these using a json-file driver for docker. All of these occurred with journald**

## Variation A: logs -f terminates early and log is permanently corrupt
The problem is reproduced with something like:
$ oc new-build ... --name=logtest
$ oc logs -f bc/logtest
```
Cloning "https://github.com/openshift/ruby-hello-world" ...
	Commit:	e79d8870be808a7abb4ab304e94c8bee69d909c6 (Merge pull request #53 from danmcp/master)
	Author:	Ben Parees <bparees@users.noreply.github.com>
	Date:	Tue Apr 5 10:06:50 2016 -0400
...
...
Successfully built ee3829579aa9
Pushing image 172.30.221.132:5000/myproject/stress:latest ...
Pushed 0/11 layers, 0% complete
Pushed 1/11 layers, 13% complete
Pushed 2/11 layers, 21% complete
Pushed 3/11 layers, 27% complete
```
**At this point, oc logs -f returns without reporting an error**

If, after the build finishes, you run oc logs bc/stress again:
```
Cloning "https://github.com/openshift/ruby-hello-world" ...
	Commit:	e79d8870be808a7abb4ab304e94c8bee69d909c6 (Merge pull request #53 from danmcp/master)
	Author:	Ben Parees <bparees@users.noreply.github.com>
	Date:	Tue Apr 5 10:06:50 2016 -0400
...
...
Successfully built ee3829579aa9
Pushing image 172.30.221.132:5000/myproject/stress:latest ...
Pushed 0/11 layers, 0% complete
Pushed 1/11 layers, 13% complete
Pushed 2/11 layers, 21% complete
Pushed 3/11 layers, 27% complete
Push successful
```
**Note the missing % complete lines and the addition of the last line, "Push successful"**

Analysis:
- Occurred approximately 30% of tests.
- Running docker logs on the resulting build container returns identical (incomplete) results. 
- This is not merely a transport issue for the logs. The logs are not intact. 
- Manually verified that the missing entries are simply not present in journald. 
- After restarting my journald service, I've been unable to recreate this. 

## Variation B: logs -f terminates prematurely but log is complete on subsequent check

This variation began after restarting journald while analyzing variation A.

$ oc new-build ... --name=logtest
$ oc logs -f bc/logtest
```
Cloning "https://github.com/openshift/ruby-hello-world" ...
	Commit:	e79d8870be808a7abb4ab304e94c8bee69d909c6 (Merge pull request #53 from danmcp/master)
	Author:	Ben Parees <bparees@users.noreply.github.com>
	Date:	Tue Apr 5 10:06:50 2016 -0400
...
...
Successfully built ee3829579aa9
Pushing image 172.30.221.132:5000/myproject/stress:latest ...
Pushed 0/11 layers, 0% complete
Pushed 1/11 layers, 13% complete
Pushed 2/11 layers, 21% complete
Pushed 3/11 layers, 27% complete
```
**At this point, oc logs -f returns without reporting an error**

If, after the build finishes, you run oc logs bc/stress again:
```
Cloning "https://github.com/openshift/ruby-hello-world" ...
	Commit:	e79d8870be808a7abb4ab304e94c8bee69d909c6 (Merge pull request #53 from danmcp/master)
	Author:	Ben Parees <bparees@users.noreply.github.com>
	Date:	Tue Apr 5 10:06:50 2016 -0400
...
...
Successfully built ee3829579aa9
Pushing image 172.30.221.132:5000/myproject/stress:latest ...
Pushed 0/11 layers, 0% complete
Pushed 1/11 layers, 13% complete
Pushed 2/11 layers, 21% complete
Pushed 3/11 layers, 27% complete
...
Pushed 8/11 layers, 73% complete
Pushed 9/11 layers, 82% complete
Pushed 10/11 layers, 91% complete
Pushed 11/11 layers, 100% complete
Push successful
```
**Logs are complete on follow-up request**


## Variation C: docker logs -f terminates prematurely right before "Push Successful", but log is complete on subsequent check

Approximately 30% of the time.

$ oc new-build ... --name=logtest
$ CID=`use a bash script to find the container id for the build container`
$ docker logs -f $CID
ex: https://gist.github.com/jupierce/2cfa7bfdc1932980ab51ae749aa09335

Outputs
```
....
Pushed 3/11 layers, 27% complete
Pushed 4/11 layers, 36% complete
Pushed 5/11 layers, 45% complete
Pushed 6/11 layers, 55% complete
Pushed 7/11 layers, 64% complete
Pushed 8/11 layers, 73% complete
Pushed 9/11 layers, 84% complete
Pushed 10/11 layers, 93% complete
Pushed 11/11 layers, 100% complete
```
**Note missing "Push Successful"**

Subsequently, docker logs $CID will show the entire log, including "Push successful"

**Comment by bparees** [comment]
can you see if any of these issues can be reproduced in a generic pod with oc logs?  (run a onetime pod that prints some output and exits, follow the logs on it, rinse/repeat)

it would be good to understanding if it's fundamental the the logs endpoint(probably is) or something in our log relaying for build logs (probably isn't).

**Comment by jupierce** [comment]
I was able to reproduce Variation C consistently without involving OpenShift, but the problem does appear to be specific to build logs. In order to create the problem, I crafted a privileged container that, as part of its entrypoint, ran a script which built & pushed a new docker image (similar to the underlying behavior of s2i). Consider the entrypoint:

```
....create new docker build with some random data to ensure push operation....
....tag the new image  to docker.io/jupierce/random....
docker push docker.io/jupierce/random:latest
echo This is a line to stderr 2>&1
```

If this script is built into a container X as its entrypoint, any lines after the push were not captured by a follow of the container logs. For example:
1. docker start X
2. docker logs -f X

In this scenario, "This is a line to stderr" is periodically not output by logs -f . A subsequent invocation of docker logs will include this line.
